### PR TITLE
[WIP, INFO] Signers

### PR DESCRIPTION
**Pull request by Mikefluff**

### CODE REVIEW

**Comment by ozalexo** [comment]
There are a lot of error found by ESlint `yarn lint:errors`
Here is the list of filtered errors (removed code-style and non-significant errors).
[Signer_LintErrors.txt](https://github.com/ChronoBank/ChronoMint/files/2314281/Signer_LintErrors.txt)

Also PR still contains obsolete code (imports of deleted files).
Could you please check?

File: packages/core/redux/device/actions.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  DEVICE_LOAD,
+  DEVICE_SET_STATUS,
+} from './constants'
+
+export const deviceAdd = (wallet) => (dispatch) => {
+  dispatch({ type: DEVICE_ADD, wallet })
+}
+
+export const deviceSelect = (wallet) => (dispatch) => {
+  dispatch({ type: DEVICE_SELECT, wallet })
```
[the remark this answers is not in the document]
Comment (wAISw) [reply packages/core/redux/device/actions.js]: `
export const deviceSelect = (wallet) => ({ type: DEVICE_SELECT, wallet })
`
that's better

File: packages/core/redux/ducks.js [hunk 2]
```diff
@@ -8,9 +8,10 @@ import { combineReducers } from 'redux-immutable'
 import assetsHolder from './assetsHolder/reducer'
 import assetsManager from './assetsManager/reducer'
 import dao from './daos/reducer'
+import device from './device/reducer'
 import ethereum from './ethereum/reducer'
 import ethMultisigWallet from './multisigWallet/reducer'
-import mainWallet from './mainWallet/reducer'
+//import mainWallet from './mainWallet/reducer'
```
Comment (wAISw) [inline packages/core/redux/ducks.js]: id useless, remove

File: packages/core/redux/ethereum/actions.js [hunk 3]
```diff
@@ -90,8 +90,11 @@ export const processTransaction = ({ web3, entry, signer }) => async (dispatch,
 }
 
 export const signTransaction = ({ entry, signer }) => async (dispatch) => {
+  console.log('sign Transaction')
```
Comment (wAISw) [inline packages/core/redux/ethereum/actions.js]: eslint error.

File: packages/core/redux/ethereum/actions.js [hunk 4]
```diff
@@ -90,8 +90,11 @@ export const processTransaction = ({ web3, entry, signer }) => async (dispatch,
...
+  console.log(entry)
```
Comment (wAISw) [inline packages/core/redux/ethereum/actions.js]: eslint error.

File: packages/core/redux/ethereum/actions.js [hunk 5]
```diff
@@ -90,8 +90,11 @@ export const processTransaction = ({ web3, entry, signer }) => async (dispatch,
...
   try {
-    const signed = await signer.signTransaction(omitBy(entry.tx, isNil))
+    console.log(entry.walletDerivedPath)
```
Comment (wAISw) [inline packages/core/redux/ethereum/actions.js]: eslint error.

File: packages/core/redux/ethereum/actions.js [hunk 6]
```diff
@@ -206,10 +209,8 @@ const acceptTransaction = (entry) => async (dispatch, getState) => {
     },
   })
   const state = getState()
-  let signer = getSigner(state)
-  if (entry.walletDerivedPath) {
-    signer = await SignerMemoryModel.fromDerivedPath({ seed: signer.privateKey, derivedPath: entry.walletDerivedPath })
-  }
+  const signer = getSigner(state)
+  console.log(signer)
```
Comment (wAISw) [inline packages/core/redux/ethereum/actions.js]: eslint error.

File: packages/core/redux/mainWallet/actions.js [hunk 7]
```diff
@@ -152,6 +149,7 @@ const handleToken = (token: TokenModel) => async (dispatch, getState) => {
           }),
         })
       } else {
+	console.log('Main Eth Wallet update balance')
```
Comment (wAISw) [inline packages/core/redux/mainWallet/actions.js]: eslint error.

File: packages/core/redux/mainWallet/actions.js [hunk 8]
```diff
@@ -211,6 +209,7 @@ const handleToken = (token: TokenModel) => async (dispatch, getState) => {
 }
 
 export const fetchTokenBalance = (token: TokenModel, account) => async (dispatch) => {
+  console.log('fetchTokenBalance')
```
Comment (wAISw) [inline packages/core/redux/mainWallet/actions.js]: eslint error.

File: packages/core/redux/mainWallet/actions.js [hunk 9]
```diff
@@ -223,18 +222,19 @@ export const fetchTokenBalance = (token: TokenModel, account) => async (dispatch
 }
 
 export const initMainWallet = () => async (dispatch) => {
+  console.log('initMainWallet')
```
Comment (wAISw) [inline packages/core/redux/mainWallet/actions.js]: eslint error.

File: packages/core/redux/mainWallet/actions.js [hunk 10]
```diff
@@ -223,18 +222,19 @@ export const fetchTokenBalance = (token: TokenModel, account) => async (dispatch
...
   dispatch({ type: WALLET_INIT, isInited: true })
 
   dispatch(subscribeOnTokens(handleToken))
 
   const providers = [
```
Comment (wAISw) [inline packages/core/redux/mainWallet/actions.js]: I suggest, refactor this code.

File: packages/core/redux/persistAccount/actions.js [hunk 11]
```diff
@@ -56,6 +54,8 @@ export const accountUpdate = (wallet) => (dispatch, getState) => {
   const state = getState()
 
   const { walletsList } = state.get(DUCK_PERSIST_ACCOUNT)
+  console.log('walletsList')
+  console.log(walletsList)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/actions.js]: eslint error.

File: packages/core/redux/persistAccount/actions.js [hunk 12]
```diff
@@ -67,11 +67,28 @@ export const accountUpdate = (wallet) => (dispatch, getState) => {
...
+ // console.log(address) 
+ // const address2 = await btcSigner.getAddress("m/44'/1'/0'/0/1")
+ // console.log(address2)
+ // console.log(btcSigner)
+//  await btcSigner.buildTx("44'/1'/0'/0/0")
+  const account = new AccountModel({
+    entry,
+    privateKey,
+  })
+
```
Comment (wAISw) [inline packages/core/redux/persistAccount/actions.js]: Fix eslint errors

File: packages/core/redux/persistAccount/actions.js [hunk 13]
```diff
@@ -99,68 +116,30 @@ export const resetPasswordAccount = (wallet, mnemonic, password) => async (dispa
...
-    encrypted: wallet && wallet.encrypt(password),
+    type,
+    encrypted: [wallet],
     profile: null,
   })
 
+  console.log(entry)
   const newAccounts = await dispatch(setProfilesForAccounts([entry]))
-
+  console.log(newAccounts)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/actions.js]: eslint error.

File: packages/core/redux/persistAccount/actions.js [hunk 14]
```diff
@@ -99,68 +116,30 @@ export const resetPasswordAccount = (wallet, mnemonic, password) => async (dispa
...
-
-  const newAccounts = await dispatch(setProfilesForAccounts([entry]))
-
-  return newAccounts[0] || entry
+export const createMemoryAccount = ({name, password, mnemonic, privateKey}) => async (dispatch) => {
+  const wallet = await EthereumMemoryDevice.create({ privateKey, mnemonic, password })
+  console.log('create account')
+  console.log(wallet)
+  const account = await dispatch(createAccount({name, wallet, type: 'memory'}))
+  console.log(account)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/actions.js]: Please, remove console.log

File: packages/core/redux/persistAccount/reducer.js [hunk 15]
```diff
@@ -13,17 +13,18 @@ const persistConfig = {
...
   selectedWallet: null,
   decryptedWallet: null,
-  rehydrated: false,
   customNetworksList: [],
 }
 
 const persistAccount = (state = initialState, action) => {
   switch (action.type) {
     case REHYDRATE:
+      console.log('rehydrated')
```
Comment (wAISw) [inline packages/core/redux/persistAccount/reducer.js]: eslint error.

File: packages/core/redux/persistAccount/reducer.js [hunk 16]
```diff
@@ -45,6 +46,8 @@ const persistAccount = (state = initialState, action) => {
       }
 
     case a.WALLETS_LOAD :
+      console.log('wallets load')
+      console.log(action.wallet)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/reducer.js]: eslint error.

File: packages/core/redux/persistAccount/selectors.js [hunk 17]
```diff
@@ -6,16 +6,48 @@
...
 
 export const getPersistAccount = (state) => {
   return state.get(DUCK_PERSIST_ACCOUNT)
 }
 
 export const getSigner = (state) => {
-  const { decryptedWallet } = getPersistAccount(state)
-  return decryptedWallet
+  const account = getPersistAccount(state)
+  console.log(account)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/selectors.js]: eslint error.

File: packages/core/redux/persistAccount/utils.js [hunk 18]
```diff
@@ -65,20 +69,21 @@ export const createAccountEntry = (name, walletFileImportObject, profile = null)
...
-    .getAddressString()
-}
-
-export const getAddressByPrivateKey = (privateKey) => {
-  return privateKeyProvider
-    .createEthereumWallet(privateKey)
-    .getAddressString()
+export const createDeviceAccountEntry = (name, device, profile = null) => {
+  console.log('create device account')
+  console.log(device)
```
Comment (wAISw) [inline packages/core/redux/persistAccount/utils.js]: eslint error.

File: packages/core/redux/tokens/actions.js [hunk 19]
```diff
@@ -166,13 +161,13 @@ export const initTokens = () => async (dispatch, getState) => {
...
     .fetchTokens()
 
-  dispatch(initBtcLikeTokens())
-  dispatch(initNemTokens())
-  dispatch(initWavesTokens())
-  dispatch(watchLatestBlock())
+//  dispatch(initBtcLikeTokens())
+//  dispatch(initNemTokens())
+//  dispatch(initWavesTokens())
+//  dispatch(watchLatestBlock())
```
Comment (wAISw) [inline packages/core/redux/tokens/actions.js]: Why you comment this code?

File: packages/core/redux/tokens/actions.js [hunk 20]
```diff
@@ -330,4 +325,6 @@ export const estimateBtcFee = (params, callback) => async () => {
   } catch (e) {
     callback(e)
   }
-}
+<<<<<<< HEAD
```
Comment (wAISw) [inline packages/core/redux/tokens/actions.js]: remove it

File: packages/core/redux/wallets/actions.js [hunk 21]
```diff
@@ -55,30 +44,25 @@ export const setWalletBalance = (walletId, balance) => (dispatch) => dispatch({
...
-    bccProvider,
-    btgProvider,
-    ltcProvider,
-    btcProvider,
-    nemProvider,
-    wavesProvider,
-    ethereumProvider,
-  ]
+const initWalletsFromKeys = () => (dispatch, getState) => {
+  console.log('Init wallets from keys')
```
Comment (wAISw) [inline packages/core/redux/wallets/actions.js]: remove console.log or add eslint ignore comment.

File: packages/core/redux/wallets_btc/actions.js [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+
+const isOwner = (wallet, account) => {
+  return wallet.owners.includes(account)
+}
+
+export const get2FAEncodedKey = (callback) => () => {
+  return ethereumProvider.get2FAEncodedKey(callback)
+}
+
+export const setWalletName = (walletId, name) => (dispatch) => dispatch({ type: WALLETS_SET_NAME, walletId, name })
```
Comment (wAISw) [inline packages/core/redux/wallets_btc/actions.js]: change to this: 
`
export const setWalletName = (walletId, name) => ({ type: WALLETS_SET_NAME, walletId, name })
`

File: packages/core/services/signers/BitcoinLedgerDevice.js [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+
+  async buildTx(path) {
+
+const network = bitcoin.networks.testnet
+
+const BLOCK_EXPLORER = axios.create({
+    baseURL: 'https://testnet.blockexplorer.com/'
+  })
+
+const MEMORY_ADDRESS = 'mreoLKdNMwnKuGq5MPjxbWzjNuojJdHB9x'
```
Comment (wAISw) [inline packages/core/services/signers/BitcoinLedgerDevice.js]: What is this code for?

File: packages/core/redux/wallets_btc/actions.js [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+          if (balance) {
+            dispatch(setWalletBalance(wallet.id, new Amount(balance, token.symbol(), true)))
+          }
+        }
+      }
+      dispatch(subscribeOnTokens(updateBalance))
+    })
+}
+
+export const subscribeWallet = ({ wallet }) => async (dispatch) => {
```
Comment (ozalexo) [inline packages/core/redux/wallets_btc/actions.js]: This method do nothing. What is should do?

**Comment by ozalexo** [comment]
Please review (and apply) my PR into Signers branch: https://github.com/Mikefluff/ChronoMint/pull/3

File: packages/core/services/signers/BitcoinTrezorDeviceMock.js [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  // this method is a part of base interface
+  getAddress (path) {
+    return this._getDerivedWallet(path).getAddress()
+  }
+
+  signTransaction (rawTx, path) { // tx object
+    const txb = new bitcoin.TransactionBuilder.fromTransaction (
+      bitcoin.Transaction.fromHex (txhex), this.network)
```
Comment (ozalexo) [inline packages/core/services/signers/BitcoinTrezorDeviceMock.js]: There is no txhex in the TxEntryModel.js. Maybe we should use TxEntryModel.hash here?

File: packages/core/services/signers/BitcoinLedgerDeviceMock.js [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  // this method is a part of base interface
+  getAddress (path) {
+    return this._getDerivedWallet(path).getAddress()
+  }
+
+  signTransaction (rawTx, path) { // tx object
+    const txb = new bitcoin.TransactionBuilder.fromTransaction (
+      bitcoin.Transaction.fromHex (txhex), this.network)
```
Comment (ozalexo) [inline packages/core/services/signers/BitcoinLedgerDeviceMock.js]: There is no txhex in the TxEntryModel.js. Maybe we should use TxEntryModel.hash here?

File: packages/core/services/signers/DashMemoryDevice.js [hunk 27]
```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright 2017–2018, LaborX PTY
+ * Licensed under the AGPL Version 3 license.
+ */
+
+import EventEmitter from 'events'
+import dashcore from 'dashcore-lib'
```
Comment (ozalexo) [inline packages/core/services/signers/DashMemoryDevice.js]: There is no dashcore-lib in the core/packages.json.

File: packages/core/redux/persistAccount/actions.js [hunk 28]
```diff
@@ -67,11 +69,28 @@ export const accountUpdate = (wallet) => (dispatch, getState) => {
...
+  console.log(entry)
+  const privateKey = EthereumMemoryDevice.decrypt({entry:entry.encrypted[0].wallet, password })
+  console.log(privateKey)
+  const btcSigner = new BitcoinLedgerDevice({ network: bitcoin.networks.testnet })
+ // const address = await btcSigner.getAddress("m/44'/1'/0'/0/0")
+ // console.log(address) 
+ // const address2 = await btcSigner.getAddress("m/44'/1'/0'/0/1")
+ // console.log(address2)
+ // console.log(btcSigner)
+  await btcSigner.buildTx("44'/1'/0'/0/0")
```
Comment (ozalexo) [inline packages/core/redux/persistAccount/actions.js]: There is no method `buildTx` in the BitcoinLedgerDevice.js

**Comment by ozalexo** [comment]
Closing. Already integrated.

**Pull request closed without merging**